policy/daemon_test.go

```diff
@@ -773,6 +778,142 @@ func Test_policyd_CheckPolicy(t *testing.T) {
+						func() *Assertion {
+							a, _ := NewAssertion("action", "domain:resource", "deny")
+							return a
+						}(),
+					})
+					return g
+				}(),
+			},
+			args: args{
+				ctx:      context.Background(),
+				domain:   "domain",
+				roles:    []string{"role1", "role2", "role3", "role4"},
+				action:   "action",
+				resource: "resource",
+			},
+			want: errors.New("policy deny: Access Check was explicitly denied"),
+		},
+		{
+			name: "check policy multiple deny deadlock",
+			fields: fields{
+				rolePolicies: func() gache.Gache {
+					g := gache.New()
+					g.Set("domain:role.role1", []*Assertion{
+						func() *Assertion {
+							a, _ := NewAssertion("action", "domain:resource", "deny")
+							return a
+						}(),
+					})
+					g.Set("domain:role.role2", []*Assertion{
+						func() *Assertion {
+							a, _ := NewAssertion("action", "domain:resource", "deny")
+							return a
+						}(),
+					})
+					g.Set("domain:role.role3", []*Assertion{
+						func() *Assertion {
+							a, _ := NewAssertion("action", "domain:resource", "deny")
+							return a
+						}(),
+					})
+					g.Set("domain:role.role4", []*Assertion{
+						func() *Assertion {
+							a, _ := NewAssertion("action", "domain:resource", "deny")
+							return a
+						}(),
+					})
+					return g
+				}(),
+			},
+			args: args{
+				ctx:      context.Background(),
+				domain:   "domain",
+				roles:    []string{"role1", "role2", "role3", "role4"},
+				action:   "action",
+				resource: "resource",
+			},
+			want: errors.New("policy deny: Access Check was explicitly denied"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			p := &policyd{
+				expireMargin:     tt.fields.expireMargin,
+				rolePolicies:     tt.fields.rolePolicies,
+				refreshDuration:  tt.fields.refreshDuration,
+				errRetryInterval: tt.fields.errRetryInterval,
+				pkp:              tt.fields.pkp,
+				etagCache:        tt.fields.etagCache,
+				etagFlushDur:     tt.fields.etagFlushDur,
+				etagExpTime:      tt.fields.etagExpTime,
+				athenzURL:        tt.fields.athenzURL,
+				athenzDomains:    tt.fields.athenzDomains,
+				client:           tt.fields.client,
+			}
+
+			b := make([]byte, 10240)
+			lenStart := runtime.Stack(b, true)
+			// t.Log(string(b[:len]))
+			err := p.CheckPolicy(tt.args.ctx, tt.args.domain, tt.args.roles, tt.args.action, tt.args.resource)
+			if err == nil {
+				if tt.want != nil {
+					t.Errorf("CheckPolicy error: err: nil, want: %v", tt.want)
+				}
+			} else {
+				if tt.want == nil {
+					t.Errorf("CheckPolicy error: err: %v, want: nil", err)
+				} else if err.Error() != tt.want.Error() {
+					t.Errorf("CheckPolicy error: err: %v, want: %v", err, tt.want)
+				}
+			}
+
+			// check runtime stack for go routine leak
+			time.Sleep(time.Second) // wait for some background process to cleanup
+			lenEnd := runtime.Stack(b, true)
+			// t.Log(string(b[:len]))
+			if math.Abs(float64(lenStart-lenEnd)) > 5 {
```

Comment:
how to define 5? if we add new feature the number maybe increase?

Reply:
this compare the runtime stack
sample: https://github.com/yahoojapan/athenz-authorizer/blob/bug/check-policy-multiple-role-deny-deadlock/output.txt

the memory address of the code may have 1 or 2 bytes difference,
but if there is a go routine leak, the difference will much larger

Reply:
I think later changes maybe lead the runtime stack changes? but I think it's ok for now~